lib/gears/surface.lua

```diff
@@ -205,6 +206,55 @@ function surface.apply_shape_bounding(draw, shape, ...)
   draw.shape_bounding = img._native
 end
 
+local function run_in_hierarchy(self, cr, width, height)
+    local function main_layout(h)
+        h:update({dpi=96}, self, width, height)
```

Comment:
Hmm. This doesn't handle redraws due to relayouts, does it? That part of the code works with a `cairo.Region` that is returned from `:update()`.